iceoryx_posh/test/moduletests/test_popo_client_port.cpp

```diff
@@ -0,0 +1,873 @@
+
+    auto responseCaproMessage = sut.portRouDi.dispatchCaProMessageAndGetPossibleResponse(caproMessage);
+
+    EXPECT_THAT(sut.portUser.getConnectionState(), Eq(iox::ConnectionState::DISCONNECT_REQUESTED));
+    ASSERT_TRUE(responseCaproMessage.has_value());
+    EXPECT_THAT(responseCaproMessage->m_serviceDescription, Eq(sut.portData.m_serviceDescription));
+    EXPECT_THAT(responseCaproMessage->m_type, Eq(iox::capro::CaproMessageType::DISCONNECT));
+    EXPECT_THAT(responseCaproMessage->m_chunkQueueData, Eq(&sut.portData.m_chunkReceiverData));
+}
+
+TEST_F(ClientPort_test, StateDisconnectRequestedWithCaProMessageTypeAckTransitionsToStateNotConnected)
+{
+    auto& sut = initAndGetClientPortForStateTransitionTests();
+    sut.portUser.connect();
+    tryAdvanceToState(sut, iox::ConnectionState::DISCONNECT_REQUESTED);
+
+    auto caproMessage = CaproMessage{CaproMessageType::ACK, sut.portData.m_serviceDescription};
+    caproMessage.m_chunkQueueData = &serverChunkQueueData;
+
+    auto responseCaproMessage = sut.portRouDi.dispatchCaProMessageAndGetPossibleResponse(caproMessage);
+
+    EXPECT_THAT(sut.portUser.getConnectionState(), Eq(iox::ConnectionState::NOT_CONNECTED));
+    ASSERT_FALSE(responseCaproMessage.has_value());
+}
+
+TEST_F(ClientPort_test, StateDisconnectRequestedWithCaProMessageTypeNackTransitionsToStateNotConnected)
+{
+    auto& sut = initAndGetClientPortForStateTransitionTests();
+    sut.portUser.connect();
+    tryAdvanceToState(sut, iox::ConnectionState::DISCONNECT_REQUESTED);
+
+    auto caproMessage = CaproMessage{CaproMessageType::NACK, sut.portData.m_serviceDescription};
+    caproMessage.m_chunkQueueData = &serverChunkQueueData;
+
+    auto responseCaproMessage = sut.portRouDi.dispatchCaProMessageAndGetPossibleResponse(caproMessage);
+
+    EXPECT_THAT(sut.portUser.getConnectionState(), Eq(iox::ConnectionState::NOT_CONNECTED));
+    ASSERT_FALSE(responseCaproMessage.has_value());
+}
+
+// END Valid transitions
+
+
+// BEGIN Invalid transitions
+
+TEST_F(ClientPort_test, InvalidStateTransitionsCallErrorHandler)
+{
+    constexpr iox::ConnectionState ALL_STATES[]{iox::ConnectionState::NOT_CONNECTED,
+                                                iox::ConnectionState::CONNECT_REQUESTED,
+                                                iox::ConnectionState::WAIT_FOR_OFFER,
+                                                iox::ConnectionState::CONNECTED,
+                                                iox::ConnectionState::DISCONNECT_REQUESTED};
+
+    // disable logging to prevent spamming the console with LogFatal outputs
+    auto logLevelScopeGuard = iox::LoggerPosh().SetLogLevelForScope(iox::log::LogLevel::kOff);
+
+    for (auto targetState : ALL_STATES)
+    {
+        for (int32_t i = 0; i < static_cast<int32_t>(CaproMessageType::MESSGAGE_TYPE_END); ++i)
+        {
+            auto caproMessageType = static_cast<CaproMessageType>(i);
+            SCOPED_TRACE(std::string("Invalid transition test from ")
```

Comment:
Nice!

Reply:
yeah, I wanted to do a typed test but that would have become a mess so I did just to loops :slightly_smiling_face: